xivo_bus/mixins.py

```diff
@@ -0,0 +1,421 @@
+                except Exception:
+                    self.log.exception(
+                        'Handler \'%s\' for event \'%s\' failed',
+                        handler.__name__,
+                        event,
+                    )
+                continue
+
+    def __extract_event(self, headers, payload):
+        event = headers.get('name', None) or headers.get('x-event', None)
+        if not event and isinstance(payload, dict):
+            event = payload.get('name', None)
+        return event
+
+    def subscribe(
+        self, event, handler, headers=None, routing_key=None, headers_match_all=True
+    ):
+        headers = self.__configure_headers(event, headers, headers_match_all)
+        binding = self.__create_binding(headers, routing_key)
+        subscription = Subscription(handler, binding)
+        with self.__lock:
+            self.__subscriptions[event].append(subscription)
+        self.log.info(
+            'Registered handler \'%s\' to event \'%s\'', handler.__name__, event
+        )
+
+    def unsubscribe(self, event, handler):
+        with self.__lock:
+            subscriptions = self.__subscriptions[event].copy()
+        try:
+            for subscription in subscriptions:
+                x_event = subscription.binding.arguments.get('x-event')
+                if subscription.handler == handler and x_event == event:
+                    with self.__lock:
+                        self.__subscriptions[event].remove(subscription)
+                    self.__remove_binding(subscription.binding)
+                    self.log.info(
+                        'Unregistered handler \'%s\' from \'%s\'',
+                        handler.__name__,
+                        event,
+                    )
+                    return True
+            return False
+        finally:
+            if not self.__subscriptions[event]:
+                with self.__lock:
+                    self.__subscriptions.pop(event)
+
+    def create_connection(self):
+        connection = self.connection.clone()
+        self.__ctrl_channel = connection.channel()
+        return connection
+
+    def get_consumers(self, Consumer, channel):
+        return [
+            Consumer(
+                queues=[self.__queue],
+                callbacks=[self.on_message_received],
+                auto_declare=True,
+            )
+        ]
+
+    def on_message_received(self, body, message):
+        headers = message.headers
+        payload = body
+        event = self.__extract_event(headers, payload)
+        try:
+            headers, payload = self.unmarshal(event, headers, payload)
+        except MiddlewareError:
+            raise
+        else:
+            self.__dispatch(event, payload, headers)
+        finally:
+            message.ack()
+
+    def on_connection_error(self, exc, interval):
+        self.__ctrl_channel = None
+        super().on_connection_error(exc, interval)
+        if self.should_stop:
```

Comment:
If this should_stop refert to the ThreadingMixin, we should find another way to emitting stop
Because it add a dependency between mixin

Reply:
You are right, we should be using events from the threading library